src/routes/jira/configuration/jira-configuration-get.ts

```diff
@@ -67,23 +86,89 @@ const getInstallation = async (subscription: Subscription, log: Logger, gitHubAp
 	}
 };
 
-interface FailedConnection {
-	id: number;
-	deleted: boolean;
-	orgName?: string;
-}
+const getConnectionsAndInstallations = async (subscriptions: Subscription[], req: Request, id: number): Promise<ConnectionsAndInstallations> => {
+	const installations = await getInstallations(subscriptions, req.log, id);
+
+	const failedConnections: FailedConnection[] = await Promise.all(
+		installations.rejected.map(async (installation) => {
+			const sub = subscriptions.find((sub: Subscription) => installation.id === sub.gitHubInstallationId);
+			const repo = sub && await RepoSyncState.findOneFromSubscription(sub);
+			return {
+				id: installation.id,
+				deleted: installation.deleted,
+				orgName: repo?.repoOwner
+			};
+		}));
+
+	const successfulConnections: SuccessfulConnection[] = installations.fulfilled
+		.map((installation) => ({
+			...installation,
+			isGlobalInstall: installation.repository_selection === "all"
+		}));
+
+	return { installations, successfulConnections, failedConnections };
+};
 
-interface SuccessfulConnection extends AppInstallation {
-	isGlobalInstall: boolean;
-}
+const JiraCloudConfiguration = async (res: Response, req: Request): Promise<object> => {
+	const { jiraHost, gitHubAppId, nonce } = res.locals;
+	const subscriptions = await Subscription.getAllForHost(jiraHost);
+	const { installations, successfulConnections, failedConnections } = await getConnectionsAndInstallations(subscriptions, req, gitHubAppId);
+
+	return {
+		host: jiraHost,
+		successfulConnections,
+		failedConnections,
+		hasConnections: !!installations.total,
+		APP_URL: process.env.APP_URL,
+		csrfToken: req.csrfToken(),
+		nonce,
+		handleNavigationClassName: "add-organization-link",
+		isGitHubCloudApp: await isGitHubCloudApp(gitHubAppId)
+	};
+};
+
+const JiraCloudAndEnterpriseConfiguration = async (res: Response, req: Request): Promise<object> => {
+	const { jiraHost, gitHubAppId, nonce } = res.locals;
+	const subscriptions = await Subscription.getAllForHost(jiraHost);
+	const { installations, successfulConnections, failedConnections } = await getConnectionsAndInstallations(subscriptions, req, gitHubAppId);
+	const gheServers = await GitHubServerApp.findForInstallationId(res.locals.installation.id) as GitHubAppServerAppWithConnections[];
+
+	// Grouping the list of servers by `gitHubBaseUrl`
+	const modifiedGHEServers = await Promise.all(chain(gheServers).groupBy("gitHubBaseUrl")
+		.map(async (value, key) => ({
+			gitHubBaseUrl: key,
+			// Fetching the subscriptions for each GH Server App
+			applications: await Promise.all(value.map(async (app) => {
+				const {
+					successfulConnections,
+					failedConnections
+				} = await getConnectionsAndInstallations(subscriptions, req, app.id);
+				app.successfulConnections = successfulConnections;
+				app.failedConnections = failedConnections;
+
+				return app;
+			}))
+		})).value());
```

Comment:
This is inefficient and might timeout or fail outright on large customers.  Instead of doing a `Promise.all` within another `Promise.all` you should just get all subscriptions based on GithubServerApps first, then modify those which will drastically reduce your DB connections and won't overload it if the customer has multiple apps with subscriptions in it.

```
const subscriptions = Subscriptions.find({
   where: {
      githubAppId: {
         [Op.in]: gheServers.map(s => s.id)
      }
   }
});
```

This is still not perfect, but it's reduced your DB calls to 2 only (O^2), one for the GHE app table, and another to the subscription table, where before you had O^n+x.

This could be reduced further to a single call once we set the DB schema correctly with foreign keys.

Reply:
The loop within the loop has been separated into 2 exact loops. The times complexity should no longer occur.